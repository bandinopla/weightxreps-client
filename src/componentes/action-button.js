import { Button, CircularProgress } from "@material-ui/core";
import { useState } from "react";


export const ActionButton = ({ children, execAction, ...rest })=>{
    const [busy, setBusy] = useState();
    const callAction = async ()=>{
        setBusy(true);
        await execAction();
        setBusy(false);
    }

    return <Button disabled={busy} onClick={ ()=>callAction() } {...rest}>
                { busy? <CircularProgress size={25}/> : children}
           </Button>;
}