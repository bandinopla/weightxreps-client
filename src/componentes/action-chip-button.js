import { Chip, CircularProgress } from "@material-ui/core";
import { useState } from "react";
 
import ErrorSnackbar from "./ErrorSnackbar";
import { parseError } from "../data/db";



/**
 * @typedef {object} ActionChipButtonParams
 * @property {Class} IconClass icono
 * @property {string} label el label del boton
 * @property {string} labelWhenSending el label del boton cuando estÃ© en estado de "busy"
 * @property {(c:(status:string)=>void)=>Promise} executeAction ejecutador de la accion
 * 
 * @param {ActionChipButtonParams} param0 
 * @returns 
 */
export const ActionChipButton = ({ IconClass, label, labelWhenSending, executeAction, size = "small" })=>{
 
    const [error, setError] = useState();
    const [busy, setBusy]   = useState(false);
    const [statusText, setStatusText] = useState(null);

    const onClick = ()=>{
        setError(null);
        setBusy(true);
        setStatusText(null)

        executeAction( status=>setStatusText(status) )
            .catch( error => setError(error) )
            .finally(()=>setBusy(false)); 
    }


 
    return <><Chip icon={busy? <CircularProgress size={10}/> : <IconClass fontSize={size}/> }
                label={busy? statusText || labelWhenSending : label} 
                variant="outlined"
                clickable
                size={size}
                onClick={onClick}
                disabled={busy}
            />
            <ErrorSnackbar trigger={ error && parseError(error) } vertical="bottom" horizontal="center"/>
            </>
};