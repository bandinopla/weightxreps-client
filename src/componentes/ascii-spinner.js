import { useEffect, useRef } from "react";



/**
 * Creates an intervall that will call callback with label + spinning ascii icon
 * @param {string} label 
 * @param {(label:string)=>void} callback 
 * @returns 
 */
export function asciiSpinnerInterval (label, callback) {
    const frames = "┤┘┴└├┌┬┐";
    var _frame = 0;
    const total = frames.length;

    return setInterval( ()=>{ 
        _frame++;
        
        callback( frames.substr(_frame%total,1)+" "+ label.substr(0, _frame%label.length)+" "+ label.substr(_frame%label.length) );

    }, 100)
}

 

/**  
 * @returns 
 */
export const AsciiSpinner = ({ label, styles })=>{ 

    const ref = useRef();

    useEffect(()=>{ 

        const anim = asciiSpinnerInterval(label, txt => { 
                                if( ref.current )
                                    ref.current.innerHTML=txt ;
                            } );

        return ()=>clearInterval(anim);
    });

    return <div ref={ref} style={{...styles}}></div>; //frames.substr(frame,1)+" "+label+".".repeat( frame%3 );
}