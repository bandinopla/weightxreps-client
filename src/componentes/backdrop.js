import Backdrop from '@material-ui/core/Backdrop';
import CircularProgress from '@material-ui/core/CircularProgress';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
  backdrop: {
    zIndex: 9999,
    color: '#fff',
  },
}));

export default function OperationBackdrop({ open }) {
  const classes = useStyles(); 
  return (
    <div> 
      <Backdrop className={classes.backdrop} open={open}>
        <CircularProgress color="inherit" />
      </Backdrop>
    </div>
  );
}