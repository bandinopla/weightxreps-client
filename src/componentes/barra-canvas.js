import { useTheme } from "@material-ui/core";
import { useLayoutEffect, useRef, useState } from "react";

/**
 * Available plates.
 * @type { [w, width, heightPercent, color][] }
 */
const normalPlates = [
    [25, 5,.9,"#ff0000"],
    [20, 5,.9,"#0033CC"],
    [15, 4,.8,"#FFCC00"],
    [10, 4,.6,"#339900"],
    [5, 3,.53,"#CCCCCC"],
    [2.5, 3,.45,"#666"],
    [1.25, 2,.25,"#000"]
]

const bigPlates = [
    [50, 6,.9,"#00ff00"],
    ...normalPlates
];

/**
 * Draws a line form x,y to tx,ty and mirrors it to the other half of the canvas using the center X as pivot.
 * @param {CanvasRenderingContext2D} ctx 
 * @param {canvasMetrics} metrics 
 */
const drawLine = (ctx, metrics, x,y,tx,ty)=>{
    var path = new Path2D();
        path.moveTo(x, y);
        path.lineTo(tx, ty); 
        path.closePath();

        ctx.stroke(path);

        // Mirror the path
        ctx.save(); // Save the current transformation matrix
        ctx.translate(metrics.halfW, metrics.halfH); // Translate to the pivot point
        ctx.scale(-1, 1); // Scale on the x-axis to mirror
        ctx.translate(-metrics.halfW, -metrics.halfH); // Translate back to the original position
        ctx.stroke(path); // Stroke the mirrored path
        ctx.restore(); // Restore the previous transformation matrix
} 

/**
 * Draws a plate
 * @param {CanvasRenderingContext2D} ctx 
 * @param {canvasMetrics} metrics 
 * @param {Array<any>} info 
 * @param {number} scale scale to apply to the width of the plate
 */
const plate = (ctx, metrics, info, scale) => {

    const width = info[1] * scale;
    const plateW = width+.5;
    const halfW = plateW/2;

    const x = metrics.plateX - halfW ;
    const T = metrics.H*info[2];

    ctx.lineWidth = width;   

    const gradient = ctx.createLinearGradient(x-3, metrics.halfH, x+6, metrics.halfH+8);
    gradient.addColorStop(0, info[3]); 
    gradient.addColorStop(1, "black");
    ctx.strokeStyle = gradient; //info[2];

    drawLine(ctx, metrics, x, (metrics.H-T)/2, x, metrics.H - (metrics.H-T)/2);
     
    return x - halfW;
}

/**
 * Draws plates on the canvas representing `w`
 * @param {CanvasRenderingContext2D} ctx 
 * @param {canvasMetrics} metrics 
 * @param {number} w 
 * @returns 
 */
const drawPlatesForWeight = (ctx, metrics, w, customBarColor ) => {

    const plates = w>=400? bigPlates : normalPlates;

    ctx.clearRect(0,0,metrics.W,metrics.H);

    // background color of the bar
    const   gradient = ctx.createLinearGradient(0, metrics.halfH-2, 0, metrics.halfH+2);
            gradient.addColorStop(0, "#ccc");  
            gradient.addColorStop(1, "#aaa");

    //#region bar
    ctx.strokeStyle = customBarColor ?? gradient; 
    ctx.lineWidth = 3;  

    drawLine(ctx, metrics, 0,metrics.halfH, metrics.halfW, metrics.halfH);//bar

    ctx.lineWidth = 6;
    drawLine(ctx, metrics, metrics.barStart, metrics.halfH-4, metrics.barStart, metrics.halfH+4);//bar
    //#endregion

    w -= 20; //bar

    if(w<0) 
    {  
        return false;
    } 
 
    
    let platesToDraw = [];

    plates.forEach( p => 
        {
        const plateW = p[0];
        let pw = plateW * 2; //one each side...
        let t = Math.floor( w / pw ); 

        if( t>0 )
        {
            for (let i = 0; i < t; i++) 
            { 
                platesToDraw.push(p);
            }
        } 

        w -= pw*t; 
    }) 


    // drawplates but modify width to make sure we fit in...
    const availSpace = metrics.plateX*.9;
    const platesWidth = platesToDraw.reduce( (t, p)=>t+p[1], 0 );
    let scale = 1;

    if( platesWidth>availSpace )
    {
        scale = availSpace / platesWidth; 
    }

    //draw
    platesToDraw.forEach(p=>{
        metrics.plateX = plate(ctx, metrics, p, scale);
    });

}

/**
 * @typedef {Object} canvasMetrics
 * @property {number} H height
 * @property {number} W width
 * @property {number} halfH H/2
 * @property {number} halfW W/2
 * @property {number} barStart position at which the portion available for plates ends.
 * @property {number} plateX current position to start putting plates
 */


export function CanvasBar({ weight, reps, FallbackTo }) {

    const ref = useRef(); 
    const [fallback, setFallback] = useState(false);
    const theme = useTheme();

    useLayoutEffect(()=>{  

        const isDarkMode = theme.palette.type=='dark';
        
        /**
         * @type {CanvasRenderingContext2D}
         */
        const ctx = ref.current?.getContext("2d");

        if( !ctx )
        {
            setFallback(true);
            return; 
        }

        

        /**
         * @type {canvasMetrics}
         */
        const metrics = {
            H: ref.current.height,
            W: ref.current.width,
            halfH: ref.current.height/2,
            halfW: ref.current.width/2,
            barStart: ref.current.width/2 * .65,  
        }

        metrics.plateX = metrics.barStart;  

        drawPlatesForWeight(ctx, metrics, weight, isDarkMode? "#666" : reps===0? "#f00" : null );

        //#region Reps Label
        const lbl = reps===0? ":(" : reps>1? "x"+reps.toString() : null;
        if( lbl )
        {
            ctx.textAlign = "center"; 
            //halfW
            ctx.font = "bold 15px arial";
            ctx.textBaseline = "middle";
  
            ctx.strokeStyle = "rgb(255 255 255 / 90%)"
            ctx.strokeText(lbl, metrics.halfW, metrics.halfH);

            ctx.fillStyle = "rgb(0 0 0)"
            ctx.fillText( lbl, metrics.halfW, metrics.halfH);  
        } 
        //#endregion 


    }, [weight, reps, theme.palette.type]); 

    if( fallback )
    {
        return <FallbackTo weight={weight} reps={reps}/>
    }

    return <canvas width="138" height="25" ref={ref} style={{maxWidth:"100%", opacity:!weight || weight<20?0.3:1}}></canvas>;
}