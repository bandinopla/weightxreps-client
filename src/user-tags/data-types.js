
import { RPEChip } from "../componentes/RPE-Chip";
import {TagBW} from "./to-view/TAG_BW";
import TAG_RANK_toView from "./to-view/TAG_RANK";

// icons
import AccessibilityNewRoundedIcon from '@material-ui/icons/AccessibilityNewRounded'; //BW
import FlashOnRoundedIcon from '@material-ui/icons/FlashOnRounded'; //FPE
import TimerRoundedIcon from '@material-ui/icons/TimerRounded'; //duration
import FavoriteRoundedIcon from '@material-ui/icons/FavoriteRounded'; //blood
import StarHalfRoundedIcon from '@material-ui/icons/StarHalfRounded'; //rank
import ShowChartRoundedIcon from '@material-ui/icons/ShowChartRounded'; //int  
import ChatBubbleIcon from '@material-ui/icons/ChatBubble';
import { UserTypedText } from "../componentes/user-typed-text";

export const TAG_PREFIX = "@";
export const TAG_START_OF_DEFINITION_REGEXP = new RegExp(`^\\s*${TAG_PREFIX}\\s*`,"i")



/**
 * TYPES is an object where each key is the ID of the type of data that a tag can hold.
 * a tag is like a variable in a programing language. 
 */
export const TYPES = {

    "TAG_BW": {
        dataTypeDesc    : "Bodyweight",
        description     : "(Automatic) Bodyweight of the day",
        value2view      : v=><TagBW value={ parseInt(v)/1000 } />, 
        icon            : AccessibilityNewRoundedIcon
    }, 
    

    "TAG_RPE": {
        reg          : /^\s*RPE\s*(1?\d(?:\.\d)?)/i,
        dataTypeDesc : "RPE",
        description  : "Tag the RPE of an arbitrary aspect of your training",
        example      : `${TAG_PREFIX} Workout Dificulty: RPE 7`,
        editor2value : m=>m[1],  
        value2view   : v=><RPEChip value={parseFloat(v)}/>, 
        value2editor : v=>`RPE ${v}`,
        icon         : FlashOnRoundedIcon
    },

    "TAG_RPE2": {
        kindOf       : "TAG_RPE",
        reg          : /^\s*(1?\d(?:\.\d)?)\s*RPE/i,
        example      : `${TAG_PREFIX} Accesory work dificulty: 7 RPE`
    } , 


    "TAG_TIME_h": {
        reg                 : /^\s*(\d+(?:\.\d+)?)\s*hs?/i,
        dataTypeDesc        : "Duration",
        icon                : TimerRoundedIcon,
        description         : "Tag the duration of something...",
        example             : `${TAG_PREFIX} Workout duration (in hours): 1.5h`,

        editor2value        : m=>m[1],
        value2number        : val=>parseFloat(val) * 3600000,   //to milliseconds
        components2value    : c => (c / 3600000).toFixed(1),
        //value2view          : v=>`${v}h`, 
        value2editor        : v=>`${v}h`,

        value2view (v) {
 
            const milliseconds      = this.value2number(v);
 
            const seconds           = Math.floor(milliseconds / 1000);
            const hours             = Math.floor(seconds / 3600);
            const minutes           = Math.floor((seconds % 3600) / 60);
            const remainingSeconds  = seconds % 60;
            
            const padZero = (value) => (value < 10 ? `0${value}` : value);
            const formattedHours = padZero(hours);
            const formattedMinutes = padZero(minutes);
            const formattedSeconds = padZero(remainingSeconds);

            if( hours<1 )
            {
                return `${formattedMinutes}:${formattedSeconds}`;
            }
            
            return `${formattedHours}:${formattedMinutes}:${formattedSeconds}`;
      
              
        }
    },

    "TAG_TIME_hm": {
        kindOf              : "TAG_TIME_h",
        //reg                 : /^\s*(?:(\d+)\s*hs?\s*)?(\d+)\s*m(?:ins?)?/i,
        reg                 : /^\s*(?:(\d+)\s*hs?)(?:\s*(\d+)\s*m(?:ins?)?)?(?:\s*(\d+(?:\.\d+)?)\s*s(?:ecs?)?)?/,
        example             : ["Session duration: 2h 25min","Time on the bike: 1h15m15sec", "Cardio: 1h 20.3s"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n") ,
        //editor2value        : m=>(m[1]??"0")+"|"+m[2]+"|0",
        editor2value        : m=> {

            const h = parseInt(m[1]);
            const min = parseInt(m[2] ?? 0);
            let s = parseFloat(m[3] ?? 0);

            const ms = Math.floor( s % 1 * 1000);
            s = Math.floor(s);

            return [h,min,s,ms].join("|"); 
        },

        value2number        : val=>TYPES.TAG_TIME_hms.value2number(val), 
        components2value    : c=>TYPES.TAG_TIME_hms.components2value(c), 

        value2editor        : val=>{
            const [ HH, MM, SS, MS ] = val.split("|").map(v=>parseInt(v));
            let sec = SS+MS/1000;

            return (HH>0? HH+"hs " : "") + (MM>0? MM+"min " : "")+(sec>0? sec+"sec " : "");
        },
    },

    "TAG_TIME_ms": {
        kindOf              : "TAG_TIME_h",
        reg                 : /^\s*(?:\s*(\d+)\s*m(?:ins?)?)(?:\s*(\d+(?:\.\d+)?)\s*s(?:ecs?)?)?/,
        example             : [ "Rest time: 3 mins 14sec", "Burpees: 2m15.2s" ].map( s=>`${TAG_PREFIX} ${s}` ).join("\n") ,
        editor2value        : m=>TYPES.TAG_TIME_hm.editor2value([null,"0", m[1], m[2]]),
        value2number        : val=>TYPES.TAG_TIME_hms.value2number(val), 
        components2value    : c=>TYPES.TAG_TIME_hms.components2value(c), 
        value2editor        : val=>{
            const [ _, M, S, MS ] = val.split("|").map(v=>parseInt(v));
            const sec = S+( MS /1000);
            return M+"min"+(sec>0? " "+sec+"sec" : "");
        },
    },

    "TAG_TIME_sec": {
        kindOf              : "TAG_TIME_h",
        reg                 : /^\s*(\d+(?:\.\d+)?)\s*s(?:ecs?)?/i,
        example             : ["3 sec", "5secs", "2.5s"].map( s=>`${TAG_PREFIX} Rest per set (in seconds): ${s}` ).join("\n") ,
        editor2value        : m=>m[1],
        value2number        : val=>parseFloat(val) * 1000,
        components2value    : c => {
            const sec = c / 1000;
            return (sec%1? sec.toFixed(1) : sec) + "s"; 
        }, 
        value2editor        : v=>`${v}`,
    },

    "TAG_TIME_hms": {
        kindOf              : "TAG_TIME_h",
        reg                 : /^\s*(?:(\d{1,3}):)?(\d{1,2}):(\d{1,2})(?:\.(\d{1,3}))?/i,
        example             : ["Rest between sets (MMSS): 3:24","Workout duration (HHMMSS): 1:20:00"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n") ,
        editor2value        : m=>[ m[1] ?? '0',m[2], m[3], m[4] ?? '0'] .join("|"),

        value2editor        : val=>{

            const [ HH, MM, SS, MS] = val.split("|");

            return `${ parseInt(HH)>0? HH.padStart(2, '0') + ":" : ""}${ MM.padStart(2, '0') }:${ SS.padStart(2, '0') }${MS?`.${MS}`:""}`;
        },
 
        value2number        : val=>{

            const [ HH, MM, SS, MS] = val.split("|").map(d=>parseInt(d));
            return HH*3600000 + MM*60000 + SS*1000 + (MS ?? 0);
        },

        components2value: milliseconds => {

            const seconds = Math.floor(milliseconds / 1000);
            const hours = Math.floor(seconds / 3600);
            const minutes = Math.floor((seconds % 3600) / 60);
            const remainingSeconds = seconds % 60;
            const _milliseconds = milliseconds % 1000;
          
            const padZero = (value) => (value < 10 ? `0${value}` : value);
            const formattedHours = padZero(hours);
            const formattedMinutes = padZero(minutes);
            const formattedSeconds = padZero(remainingSeconds);
          
            return `${formattedHours}|${formattedMinutes}|${formattedSeconds}|${_milliseconds}`;
        }
    },


    //MAP = (2/3) * diastolic pressure + (1/3) * systolic pressure
    "TAG_BLOOD": {
        reg          : /^\s*(\d+)\s*-\s*(\d+)/i,
        icon         : FavoriteRoundedIcon,
        dataTypeDesc : "Blood Pressure",
        description  : "Track blood pressure values (diastolic-systolic)",
        example      :  ["Blood Pressure: 120-80","Blood Pressure (short format): 12-8"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n") , 

        editor2value : m=>{

            let [_, systolic, diastolic ] = m; 
 

            //
            // check normal(?) values
            //
            return [['systolic', parseInt(systolic), 200, 90],['diastolic',parseInt(diastolic), 120, 60]]
            
            .map( (limit, i, arr)=>{

                let v                   = limit[1];
                const usingShortVersion = arr[0][1]<20 || arr[1][1]<20;

                if( usingShortVersion )
                {
                    //
                    // it is used the short format
                    //
                    v *= 10; 
                }

                if( v<limit[3] || v>limit[2] )
                {
                    throw new Error(`Invalid ${limit[0]}, expected a value between ${limit[3]}mmHg and ${limit[2]} mmHg, got: ${v}mmHg`);
                }

                return v;

            })

            .join("-")
            
            ; 

        },

        /**
         * Calculates the mean arterial pressure (MAP)
         */
        value2number : val=>{

            const [systolic, diastolic] = val.split("-").map(v=>parseInt(v));

            const MAP = (2/3) * diastolic + (1/3) * systolic;

            return MAP*1000000 + diastolic*1000 + systolic;

        }, 

        value2components: val => {
            return val.split("-").map(v=>parseInt(v));
        },

        components2value: comp => `${comp[0]}-${comp[1]}`,

        addComponents       : (A,B)=>{
            if( isNaN(B) )
            {
                return [ A[0]+B[0], A[1]+B[1] ]
            }
            
            return [ A[0]+B, A[1]+B ]
        },

        subComponents       : (A,B)=>{
            if( isNaN(B) )
            {
                return [ A[0]-B[0], A[1]-B[1] ]
            }
            
            return [ A[0]-B, A[1]-B ]
        },

        divComponents       : (A,B)=>{
            if( isNaN(B) )
            {
                return [ A[0]/B[0], A[1]/B[1] ]
            }
            
            return [ A[0]/B, A[1]/B ]
        },
    },  

    
    "TAG_RANK": {
        reg          : /^\s*\d+(?:\.\d)?\/\s*\d+/,
        dataTypeDesc : "Rank",
        icon         : StarHalfRoundedIcon,
        description  : "Rank a particular aspect on an arbitraty scale",
        example      : ["Stress: 4.5/10","Pain: 5/10", "Will to live: 50/100"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n"),
        editor2value : m=>m[0].replace(/\s/,""), 
        value2number : v => {
            const [ rank, outOf ] = v.split("/").map(l=>parseInt(l));
            return rank / outOf;
        },
        components2value: c => {
            let outOf = 10;
            return `${ (c*outOf).toFixed(1) }/${outOf}`;
        },
        value2view(v) { 
            return <TAG_RANK_toView value={ this.value2number(v) }/> 
        }
    }, 


    "TAG_INT": {
        reg          : /^\s*(\d+(?:\.\d+)?)/,
        dataTypeDesc : "Number",
        icon         : ShowChartRoundedIcon,
        description  : "Tag a numeric value",
        example      : ["Steps: 65","Liters of water: 3.2"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n"),
        editor2value : m=>m[1], 
    },

    "TAG_STRING": {
        reg             : /^(.*)/i,
        example         : ["Feel like: shit","mood: grumpy"].map( s=>`${TAG_PREFIX} ${s}` ).join("\n"), 
        dataTypeDesc    : "Short Text",
        icon            : ChatBubbleIcon,
        isJustText      : true, //flag to know if this tag represents arbitrary text.
        canBeCharted    : false,
        description     : "Tag a an arbitrary aspect. 12 Characters max. The graph will be flat and will only serve to show the value.",
        editor2value    : m=>{

            const v = m[1].trim();
            if( v.length>12 )
            {
                throw new Error("Max length of a tag value is 12 characters. You typed: "+v.length);
            }
            return v;

        },
        value2editor    : v=>v,
        value2view      : v=><UserTypedText text={v} noWrap/>
    },


}




export const getUserAvailableTagTypes = ()=>Object.keys(TYPES).filter( key=>!!TYPES[key].reg);
export const getUserAvailableTypes = ()=>getUserAvailableTagTypes().map(key=>TYPES[key]);

//
// default behaviour if the value is a number and it can be used for sorting and such...
// if "editor2value" returns a number, then all these default work fine.
//
const TYPE_DEFAULTS = {
    canBeCharted        : true,
    value2number        : v=>Number(v),         // a number that can be plotted on a line graph
    //value2components    : v=>Number(v),       // only define this is "value2number" would not be a valid component.
    value2view          : v=><span>{v}</span>,  // returns the visual represenation of the value so the user van view it on screen
    value2editor        : v=>`${v}`,            // converts the values back into the format that the user types on the journal editor.

    
    addComponents       : (A,B)=>A+B,
    subComponents       : (A,B)=>A-B,
    divComponents       : (A,B)=>A/B,
    components2value    : C=>String(C),         // based on the components that this type creates per value, return the value that represents this component/s
}


//
// to feed tooltips
//
const typedesc2description = new Map();


//
// copy all props of "kingOf"
//
Object.keys(TYPES).forEach(key=>{

    if( key.length>12 ) {
        throw new Error("Hey! Key ["+key+"] must be 12 characters or less!! this is because the DB stores this in a VARCHAR(12)")
    }

    TYPES[key].key = key;

    if( TYPES[key].kindOf )
    {
        TYPES[key] = {
            ...TYPES[TYPES[key].kindOf],
            ...TYPES[key], 
        }
    }
    else 
    {
        TYPES[key] = {
            ...TYPE_DEFAULTS,
            ...TYPES[key], 
        }
    }

    typedesc2description.set( TYPES[key].dataTypeDesc, TYPES[key].description );
})


/**
 * returns the description of the TYPE with this type as "dataTypeDesc"
 */
export const tagTypeDesc2Description = type => {
    return typedesc2description.get( type );
}