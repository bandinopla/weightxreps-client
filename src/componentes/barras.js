import flags_img_src from "./barras.png";
import "./barras.css";
import FireIcon from "../codemirror/fire.png";
import { CanvasBar } from "./barra-canvas";

function BarraOld({ weight, reps }) {

    var n = Math.round( (weight - 20) / 2.5 );

    if( weight>500 )
    {
        return <><img src={FireIcon} style={{maxWidth:12}}/> <strong>A LOT</strong> <img src={FireIcon}style={{maxWidth:12}}/></>;
    }

    return <i className={"bar i"+n} style={{ backgroundImage:"url("+flags_img_src+")", opacity:weight<20?0.5:1 }}>
        { reps && <b style={{backgroundColor:"rgba(0,0,0,0.5)", color:"white", padding:4}}>x{reps}</b>}</i>
}  
 

export default function Barra({ weight, reps }) { 

    return <CanvasBar weight={weight} reps={reps} FallbackTo={BarraOld} > 
    </CanvasBar>
}