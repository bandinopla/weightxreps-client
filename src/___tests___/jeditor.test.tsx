import React from 'react';
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'

import { ActionButton } from '../componentes/action-button';
import { render, screen } from '@testing-library/react';
import { JLogTokenizer } from '../componentes/journal/tokenizer';
import { TYPES as TagTypes } from '../user-tags/data-types';

// it('renders without crashing', () => {
//   const div = document.createElement('div');

//   createRoot(div)
//         .render(<div><ActionButton execAction={()=>{}}>aaaaa</ActionButton></div> );
// });

beforeEach(() => {
    //jest.spyOn(console, 'warn').mockImplementation(() => {});
    jest.spyOn(console, 'error').mockImplementation((e) => {
        if( e.startsWith("Warning")){
             
        }
        else 
        {
            console.error(e)
        }
    });
  });
   

it('should parse a workout correctly...', async () => {
    // render(<ActionButton execAction={()=>new Promise(()=>{})}>Hola</ActionButton>);
 
    var tokenizer = await JLogTokenizer({
        tags:["Walking","Running"],
        utags:[],
        exercises:[
            { id:1, name:"Squat" },
            { id:2, name:"Bench Press"}
        ], 
        defaultYMD:"2020-01-01",
        printAllErrorsOnThrow: true
    }).textToPayload(`2020-01-12
@ 75 BW

${ Object.entries(TagTypes).map( entry=>entry[1].example ? Array.isArray(entry[1].example)? entry[1].example.join("\n") : entry[1].example :`` ).join("\n") }
Today was a good day
#Squat
60x3
100x5x5 haha
Then in the afternoon!...
#Bench Press
100x5x4 @5.5
The i just ran...
asdasd
#Farmer walk
140kg x 5m
#Running 
1300m in 10min

2023-11-12
Thenext day rest`,10, undefined); 
 
       
    expect( JSON.stringify(tokenizer) ).toMatchSnapshot();

    // await userEvent.click(screen.getByRole('button'))
    // await screen.findByRole('loader')

    // expect(screen.getByRole("loader")).toBeInTheDocument();
    // expect(1+1).toBe(3)
});