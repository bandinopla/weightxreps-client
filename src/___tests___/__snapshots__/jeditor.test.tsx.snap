// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse a workout correctly... 1`] = `"[{\\"newExercise\\":\\"#Farmer walk\\"},{\\"newExercise\\":\\"#Running \\"},{\\"on\\":\\"2020-01-12\\"},{\\"bw\\":75,\\"lb\\":null},{\\"tag\\":\\"Workout Dificulty\\",\\"value\\":\\"7\\",\\"type\\":\\"TAG_RPE\\"},{\\"tag\\":\\"Accesory work dificulty\\",\\"value\\":\\"7\\",\\"type\\":\\"TAG_RPE2\\"},{\\"tag\\":\\"Workout duration (in hours)\\",\\"value\\":\\"1.5\\",\\"type\\":\\"TAG_TIME_h\\"},{\\"tag\\":\\"Session duration\\",\\"value\\":\\"2|25|0|0\\",\\"type\\":\\"TAG_TIME_hm\\"},{\\"tag\\":\\"Time on the bike\\",\\"value\\":\\"1|15|15|0\\",\\"type\\":\\"TAG_TIME_hm\\"},{\\"tag\\":\\"Cardio\\",\\"value\\":\\"1|0|20|300\\",\\"type\\":\\"TAG_TIME_hm\\"},{\\"tag\\":\\"Rest time\\",\\"value\\":\\"0|3|14|0\\",\\"type\\":\\"TAG_TIME_ms\\"},{\\"tag\\":\\"Burpees\\",\\"value\\":\\"0|2|15|199\\",\\"type\\":\\"TAG_TIME_ms\\"},{\\"tag\\":\\"Rest per set (in seconds)\\",\\"value\\":\\"3\\",\\"type\\":\\"TAG_TIME_sec\\"},{\\"tag\\":\\"Rest per set (in seconds)\\",\\"value\\":\\"5\\",\\"type\\":\\"TAG_TIME_sec\\"},{\\"tag\\":\\"Rest per set (in seconds)\\",\\"value\\":\\"2.5\\",\\"type\\":\\"TAG_TIME_sec\\"},{\\"tag\\":\\"Rest between sets (MMSS)\\",\\"value\\":\\"0|3|24|0\\",\\"type\\":\\"TAG_TIME_hms\\"},{\\"tag\\":\\"Workout duration (HHMMSS)\\",\\"value\\":\\"1|20|00|0\\",\\"type\\":\\"TAG_TIME_hms\\"},{\\"tag\\":\\"Blood Pressure\\",\\"value\\":\\"120-80\\",\\"type\\":\\"TAG_BLOOD\\"},{\\"tag\\":\\"Blood Pressure (short format)\\",\\"value\\":\\"120-80\\",\\"type\\":\\"TAG_BLOOD\\"},{\\"tag\\":\\"Stress\\",\\"value\\":\\"4.5/10\\",\\"type\\":\\"TAG_RANK\\"},{\\"tag\\":\\"Pain\\",\\"value\\":\\"5/10\\",\\"type\\":\\"TAG_RANK\\"},{\\"tag\\":\\"Will to live\\",\\"value\\":\\"50/100\\",\\"type\\":\\"TAG_RANK\\"},{\\"tag\\":\\"Steps\\",\\"value\\":\\"65\\",\\"type\\":\\"TAG_INT\\"},{\\"tag\\":\\"Liters of water\\",\\"value\\":\\"3.2\\",\\"type\\":\\"TAG_INT\\"},{\\"tag\\":\\"Feel like\\",\\"value\\":\\"shit\\",\\"type\\":\\"TAG_STRING\\"},{\\"tag\\":\\"mood\\",\\"value\\":\\"grumpy\\",\\"type\\":\\"TAG_STRING\\"},{\\"text\\":\\"Today was a good day\\"},{\\"eid\\":1,\\"erows\\":[{\\"erow\\":true,\\"w\\":{\\"v\\":60},\\"r\\":3,\\"s\\":1,\\"t\\":0,\\"c\\":\\"\\",\\"d\\":null,\\"type\\":0,\\"rpe\\":null,\\"line\\":28},{\\"erow\\":true,\\"w\\":{\\"v\\":100},\\"r\\":5,\\"s\\":5,\\"t\\":0,\\"c\\":\\"haha\\",\\"d\\":null,\\"type\\":0,\\"rpe\\":null,\\"line\\":29}]},{\\"text\\":\\"Then in the afternoon!...\\"},{\\"eid\\":2,\\"erows\\":[{\\"erow\\":true,\\"w\\":{\\"v\\":100},\\"r\\":5,\\"s\\":4,\\"t\\":0,\\"c\\":\\"\\",\\"d\\":null,\\"type\\":0,\\"rpe\\":5.5,\\"line\\":32}]},{\\"text\\":\\"The i just ran...\\\\nasdasd\\"},{\\"eid\\":0,\\"erows\\":[{\\"erow\\":true,\\"w\\":{\\"v\\":140},\\"r\\":1,\\"s\\":1,\\"t\\":0,\\"c\\":\\"\\",\\"d\\":{\\"val\\":50000,\\"unit\\":\\"m\\"},\\"type\\":2,\\"rpe\\":null,\\"line\\":36}]},{\\"eid\\":-1,\\"erows\\":[{\\"erow\\":true,\\"w\\":0,\\"r\\":1,\\"s\\":1,\\"t\\":600000,\\"c\\":\\"\\",\\"d\\":{\\"val\\":13000000,\\"unit\\":\\"m\\"},\\"type\\":2,\\"rpe\\":null,\\"line\\":38}]},{\\"on\\":\\"2023-11-12\\"},{\\"text\\":\\"Thenext day rest\\"}]"`;
